### PR TITLE
Implement storage package

### PR DESCRIPTION
**Pull request by sonmezonur**

### CODE REVIEW

File: e2e/rm_test.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func TestRemoveMultipleS3Objects(t *testing.T) {
 	result.Assert(t, icmd.Success)
 
 	assertLines(t, result.Stderr(), map[int]compareFunc{
-		0: suffix(` +OK "rm s3://%v/*" (4)`, bucket),
```
Comment (igungor) [inline e2e/rm_test.go]: Have you found the cause of this change?

File: core/worker.go [hunk 2]
```diff
@@ -30,47 +27,57 @@ type WorkerPoolParams struct {
...
-	awsSession    *session.Session
-	cancelFunc    context.CancelFunc
-	st            *stats.Stats
-	idlingCounter int32
+	ctx            context.Context
+	params         *WorkerPoolParams
+	jobQueue       chan *Job
+	subJobQueue    chan *Job
+	wg             *sync.WaitGroup
+	storageFactory func() (storage.Storage, error)
```
Comment (igungor) [inline core/worker.go]: i think `newClient` sounds better

Comment (igungor) [reply core/worker.go]: or just `client()`

File: core/worker.go [hunk 3]
```diff
@@ -132,54 +128,38 @@ func (p *WorkerPool) runWorker(st *stats.Stats, idlingCounter *int32, id int) {
...
-
 		case <-p.ctx.Done():
 			return
 		}
 	}
 }
 
+// runJob runs parent job and returns sub-job.
+func (p *WorkerPool) runJob(wp WorkerParams, job *Job) *Job {
+	if err := job.Run(&wp); err != nil {
```
Comment (igungor) [inline core/worker.go]: let's handle the non-error case first to reduce the indentation.

File: core/worker.go (hunk 2 above)
Comment (sonmezonur) [reply core/worker.go]: Changed to `newClient`

File: core/worker.go (hunk 3 above)
Comment (sonmezonur) [reply core/worker.go]: Logic was moved into `job.go` and checks non-error case first.

File: storage/s3.go [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+		return err
+	}
+
+	for _, d := range o.Deleted {
+		s.stats.put(aws.StringValue(d.Key), StatsResponse{Success: true})
+	}
+
+	for _, e := range o.Errors {
+		s.stats.put(aws.StringValue(e.Key), StatsResponse{
+			Success: true,
```
Comment (igungor) [inline storage/s3.go]: should be false.

Comment (sonmezonur) [reply storage/s3.go]: done

**Pull request merged**